lyse/__init__.py

```diff
@@ -296,6 +299,102 @@ def h5_path(self):
     def no_write(self):
         """bool: The value provided for `no_write` during instantiation."""
         return self.__no_write
+    
+    @property
+    def h5_file(self):
+        """h5py.File: opened h5py file handle for the shot"""
+        return self.__h5_file
+    
+    @contextlib.contextmanager
+    def open(self, mode):
+        """Opens the Run's h5 file for successive reads/writes.
+
+        Intended to be used as a context manager.
+        Supports all h5 modes, though only `'r'` and `'r+'`/`'a'` are typically
+        used within lyse itself.
+
+        Args:
+            mode (str): which mode to open the h5 file with.
+                Must be 'r', 'a', 'r+', 'w', 'w-', or 'x'.
+                Lyse typically only uses 'r' and 'a'.
+
+        Examples:
+            >>> from lyse import *
+            >>> shot = Run(path)
+            >>> with shot.open('a'):
+            >>>     # shot processing that requires reads/writes
+            >>>     t, vals = shot.get_trace('my_trace')
+            >>>     results = vals**2
+            >>>     shot.save_result_array('my_result', results)
+            >>> # Shot processing that doesn't require h5 reads/writes
+            >>> print(f'Max of results is {results.max():.3f}')
+            Max of results is 5.003
+
+        Raises:
+            PermissionError: If the Run is set as read-only but a write mode is requested
+        """
+
+        # ensure no_write is respected
+        if self.no_write:
+            if mode == 'r':
+                pass
+            else:
+                msg = f'Cannot open file with a write mode; this run is read-only'
+                raise PermissionError(msg)
+            
+        # ensure file is not already opened
+        if self.__h5_file is not None:
+            msg = 'h5_file already opened'
+            raise PermissionError(msg)
+        
+        self.__h5_file = h5py.File(self.h5_path, mode)
+        try:
+            yield None
+        finally:
+            self.__h5_file.close()
+            self.__h5_file = None
+            
+    def open_file(mode):
```

Comment:
I think this decorator doesn’t need to be inside the `Run` class? Probably clearer if it’s not (otherwise it’s odd that it’s not defined as a `staticmethod` and also doesn’t take `self` as an argument)

Reply:
I thought that would be the case, just wanted to get another opinion.

Now that I look at it, I'm pleasantly surprised it even works as is. It's not like I called it as a typical class method. TIL more about what using self actually does.

Reply:
Ah, should have tested this out before pushing. Naive putting the decorator outside the class [doesn't work](https://stackoverflow.com/questions/1263451/python-decorators-in-classes) because you need to access the class instance from within the wrapper and that doesn't work unless we define within the class. Not sure how much I trust a single SO post for best practices, but what I had appears to be a fairly common pattern. I'll revert for now.

Reply:
This really should have worked. `self` isn't accessed until the method is called (e.g. it's used inside `wraps` which is the replacement method returned by the decorator). It's equivalent to the `define_state` decorator in BLACS. The only difference is the use of the `functools.wraps`. that should be fine butmaybe I'm missing something. Is the use of wraps meant to be on the middle function instead? It's primary purpose is to copy the docstring of the `Run` methods decorated with `open`. Should be easy enough to reason out where it needs to be and test it's copying the doc strings

Reply:
Oh, I see. Given you don't need to write to that attribute inside the decorator, can you just switch it to using the `self.h5_file` property instead? Then the decorator could live outside of the class.

Reply:
OK, latest push should resolve this one too.

Reply:
For what it's worth, the name mangling is achieving nothing and could be removed. The purpose of double-underscore mangling is to make attributes inaccessible from subclasses or parent classes, which we are not trying to do.

There's not even much reason to make them single-underscore private, Id' argue since they all have getters (I guess some don't have setters). My guess is that the main reason they are private at all is to make it easier to document them by adding docstrings to the getters.

But yeah, single-underscore is arguable, whereas double is achieving nothing.

Reply:
The name mangling, and exposure via property getters, is to prevent end users from changing the internal state. A lot of the checks, etc. rely on the assumption that these variables are private. If one of those mangled variables was changed (accidentally) I think you'd see very odd behaviour.

Reply:
Well that's an argument for them being private generally (i.e. one leading underscore), which I lean against more than you but accept there's a case for. Name mangling as triggered by two leading underscores is for privacy from methods of subclasses and parent classes though, which I don't think there is a case for.